### PR TITLE
OEM: ensure only one kernel gets installed

### PR DESCRIPTION
**Pull request by ogayot**
Some more testing is needed but it looks mostly ok.

Depends: #1702

### CODE REVIEW

**Review by dbungert — APPROVED** [review]
LGTM, minus the bit already raised on the other PR.

File: subiquity/server/controllers/install.py [hunk 1]
```diff
@@ -336,6 +350,14 @@ async def run_curtin_step(name, stages, step_config, source=None):
                 for pkg in self.model.oem.metapkgs:
                     await self.install_package(package=pkg.name)
 
+            # If we already have the needed kernel installed, don't bother
+            # requesting curthooks to install it or we might end up with two
+            # kernels.
+            for pkg in await list_installed_kernels(Path(self.tpath())):
+                if pkg == self.model.kernel.needed_kernel:
+                    self.model.kernel.curthooks_no_install
```
Comment (mwhudson) [inline subiquity/server/controllers/install.py:358]: Um I think this line is incomplete :-)

File: subiquity/server/kernel.py [hunk 2]
```diff
@@ -27,3 +32,30 @@ def flavor_to_pkgname(flavor: str, *, dry_run: bool) -> str:
     # that's a bit tricky until we get cleverer about
     # the apt config in general.
     return f'linux-{flavor}-{release}'
+
+
+async def list_installed_kernels(
+        rootfs: pathlib.Path,
+        *, grep_status: str = "grep-status") -> List[str]:
```
Comment (mwhudson) [inline subiquity/server/kernel.py:39]: This argument is unused.

File: snapcraft.yaml [hunk 3]
```diff
@@ -110,6 +110,7 @@ parts:
       # This list includes the dependencies for curtin and probert as well,
       # there doesn't seem to be any real benefit to listing them separately.
       - cloud-init
+      - dctrl-tools
```
Comment (mwhudson) [inline snapcraft.yaml:113]: I guess we could just stage /usr/bin/grep-dctrl but the whole package is fairly tiny.

File: subiquity/server/controllers/install.py [hunk 4]
```diff
@@ -336,6 +350,14 @@ async def run_curtin_step(name, stages, step_config, source=None):
                 for pkg in self.model.oem.metapkgs:
                     await self.install_package(package=pkg.name)
 
+            # If we already have the needed kernel installed, don't bother
+            # requesting curthooks to install it or we might end up with two
+            # kernels.
+            for pkg in await list_installed_kernels(Path(self.tpath())):
+                if pkg == self.model.kernel.needed_kernel:
```
Comment (mwhudson) [inline subiquity/server/controllers/install.py:357]: This is all a bit confusing but I don't think we should check this, for two reasons. One is that pkg here will be something like "linux-image-6.2.0-23-generic" or "linux-image-6.1.0-1012-oem" and I think needed_kernel is going to be a metapackage name like "linux-oem-22.04" or "linux" and the other is that even if we think we want "linux-oem-22.04" installing the metapackage might actually install "linux-oem-22.04d" or something. 

Basically IMO we should instruct curtin to not install a kernel iff there is any kernel installed already. Luckily this makes the code simpler :-)

Comment (mwhudson) [reply subiquity/server/controllers/install.py:357]: I'm not sure how things should interact if the autoinstall explicitly requests a different flavour though.

**Review by mwhudson — CHANGES_REQUESTED** [review]

File: subiquity/server/controllers/install.py [hunk 5]
```diff
@@ -336,6 +350,14 @@ async def run_curtin_step(name, stages, step_config, source=None):
                 for pkg in self.model.oem.metapkgs:
                     await self.install_package(package=pkg.name)
 
+            # If we already have the needed kernel installed, don't bother
+            # requesting curthooks to install it or we might end up with two
+            # kernels.
+            for pkg in await list_installed_kernels(Path(self.tpath())):
```
[the remark this answers is not in the document]
Comment (ogayot) [reply subiquity/server/controllers/install.py:356]: I think that would only happen on broken or half broken scenarios, like:
* two suggested OEM meta-packages pulling two different kernel flavors (for now there should be only one meta-package per hardware model - in theory)
* one kernel in the squashfs (don't think we'll ever do that) and the OEM metapkg pulling another flavor

File: subiquity/server/controllers/install.py (hunk 1 above)
Comment (ogayot) [reply subiquity/server/controllers/install.py:358]: Ouch! Thanks for noticing!

File: snapcraft.yaml (hunk 3 above)
Comment (ogayot) [reply snapcraft.yaml:113]: Yes - and no Depends so I guess I'll keep the whole package.

File: subiquity/server/kernel.py (hunk 2 above)
Comment (ogayot) [reply subiquity/server/kernel.py:39]: Oops!

File: subiquity/server/controllers/install.py (hunk 4 above)
Comment (ogayot) [reply subiquity/server/controllers/install.py:357]: Yeah, I think you're right. Dan was suggesting a runtime error if there's an explicit kernel flavor + an OEM meta-package. I think it is a sensible approach.

File: subiquity/server/controllers/install.py [hunk 6]
```diff
@@ -312,6 +315,17 @@ async def run_curtin_step(name, stages, step_config, source=None):
                 step_config=self.generic_config(),
                 source=source,
                 )
+            if self.app.opts.dry_run:
+                # In dry-run, extract does not do anything. Let's create what's
+                # needed manually. Ideally, we would not hardcode
+                # var/lib/dpkg/status because it is an implementation detail.
+                status = Path("var/lib/dpkg/status")
+                (Path(root) / status).parent.mkdir(parents=True, exist_ok=True)
```
Comment (mwhudson) [inline subiquity/server/controllers/install.py:323]: This is nitpicky to the max but a) root is already a Path here b) Path / str works fine? so you can leave status as a string.

File: subiquity/server/kernel.py (hunk 2 above)
Comment (mwhudson) [reply subiquity/server/kernel.py:39]: I admit I expected you to fix this by removing the argument :-) What is the use case for passing a different value here?

**Review by mwhudson — APPROVED** [review]
Thanks, this is good stuff. I still have a couple of comments but they are very much at the nitpicking level now.

File: subiquity/server/controllers/install.py (hunk 6 above)
Comment (ogayot) [reply subiquity/server/controllers/install.py]: Good point, updated :)

**Comment by ogayot** [comment]
> I admit I expected you to fix this by removing the argument :-) What is the use case for passing a different value here?

It felt like it could be a way to substitute a command by another when doing dry-runs - as in
```python
list_installed_kernels(..., grep_status='scripts/replay-grep-status')
```
or something.

But it's not really usable as is - so I guess it should go (and is now gone)
